### PR TITLE
[gRPC] Code refactoring for gRPC's flatbuffer support @open sesame 12/02 12:32

### PR DESCRIPTION
**Pull request by dongju-chae**
This patch performs code refactoring for gRPC's flatbuffer support.
The existing gRPC codes can be re-used for flatbuffer as well.

It also checks the same gRPC testcases for both IDLs (protobuf and flatbuf).

It resolves https://github.com/nnstreamer/nnstreamer/issues/2625.

Signed-off-by: Dongju Chae <dongju.chae@samsung.com>

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed []Skipped
2. Run test: [*]Passed [ ]Failed []Skipped

### CODE REVIEW

File: ext/nnstreamer/extra/nnstreamer_grpc_protobuf.cc [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+  *buffer = gst_buffer_new ();
+
+  for (guint i = 0; i < num_tensor; i++) {
+    const Tensor * tensor = &tensors.tensor (i);
+    const void * data = tensor->data ().c_str ();
+    gsize size = tensor->data ().length ();
+
+    memory = gst_memory_new_wrapped (GST_MEMORY_FLAG_READONLY,
+        g_memdup (data, size), size, 0, size, NULL, NULL);
```
Comment (jaeyun-jung) [inline ext/nnstreamer/extra/nnstreamer_grpc_protobuf.cc:178]: this needs g_free to release mem block:
gpointer new_mem = g_memdup (data, size);
gst_memory_new_wrapped (..., new_mem, g_free)

File: ext/nnstreamer/extra/nnstreamer_grpc_flatbuf.cc [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+  GstMemory *memory;
+
+  *buffer = gst_buffer_new ();
+
+  for (guint i = 0; i < num_tensor; i++) {
+    const Tensor * tensor = tensors->tensor ()->Get (i);
+    const void * data = tensor->data ()->data ();
+    gsize size = VectorLength (tensor->data ());
+
+    memory = gst_memory_new_wrapped (GST_MEMORY_FLAG_READONLY,
```
Comment (jaeyun-jung) [inline ext/nnstreamer/extra/nnstreamer_grpc_flatbuf.cc:198]: also this requires g_free to release newly allocated mem block

File: ext/nnstreamer/extra/nnstreamer_grpc_protobuf.cc (hunk 1 above)
Comment (dongju-chae) [reply ext/nnstreamer/extra/nnstreamer_grpc_protobuf.cc:178]: Oh, thanks. I missed to handle this. In the original codes, the callers had managed its de-allocation, but now it's no longer valid.

File: ext/nnstreamer/extra/nnstreamer_grpc_flatbuf.cc (hunk 2 above)
Comment (dongju-chae) [reply ext/nnstreamer/extra/nnstreamer_grpc_flatbuf.cc:198]: Updated

File: ext/nnstreamer/extra/nnstreamer_grpc_protobuf.cc [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  GstDataQueueItem *item;
+
+  /* until flushing */
+  while (gst_data_queue_pop (queue_, &item)) {
+    Tensors tensors;
+
+    _get_tensors_from_buffer (GST_BUFFER (item->object), tensors);
+    writer->Write (tensors);
+
+    g_free (item);
```
Comment (jaeyun-jung) [inline ext/nnstreamer/extra/nnstreamer_grpc_protobuf.cc:73]: for safe release
GDestroyNotify destroy = (item->destroy) ? item->destroy : g_free;
destroy (item);

File: ext/nnstreamer/extra/nnstreamer_grpc_flatbuf.cc [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+  GstDataQueueItem *item;
+
+  /* until flushing */
+  while (gst_data_queue_pop (queue_, &item)) {
+    Message<Tensors> tensors;
+
+    _get_tensors_from_buffer (GST_BUFFER (item->object), tensors);
+    writer->Write (tensors);
+
+    g_free (item);
```
Comment (jaeyun-jung) [inline ext/nnstreamer/extra/nnstreamer_grpc_flatbuf.cc:101]: also this line should be fixed

File: ext/nnstreamer/extra/nnstreamer_grpc_protobuf.cc (hunk 3 above)
Comment (dongju-chae) [reply ext/nnstreamer/extra/nnstreamer_grpc_protobuf.cc:73]: Thanks, updated.

File: ext/nnstreamer/extra/nnstreamer_grpc_flatbuf.cc (hunk 4 above)
Comment (dongju-chae) [reply ext/nnstreamer/extra/nnstreamer_grpc_flatbuf.cc:101]: Thanks, updated.

**Review by jaeyun-jung — APPROVED** [review]

**Review by anyj0527 — APPROVED** [review]
LGTM Thanks!

File: ext/nnstreamer/extra/meson.build [hunk 5]
```diff
@@ -32,8 +32,12 @@ if grpc_support_is_available
     error('failed to resolve dependency: gRPC requires libprotobuf.')
   endif
 
-  # Don't generate .proto file twice
-  nns_protobuf_grpc_lib = shared_library ('nnstreamer_protobuf_grpc',
+  if not flatbuf_support_is_available
+    error('failed to resolve dependency: gRPC requires libflatbuf.')
+  endif
```
Comment (myungjoo) [inline ext/nnstreamer/extra/meson.build:37]: Do you really need to make gRPC require both protobuf & flatbuf?
Or would it be possible to require at least one of the two?

Comment (dongju-chae) [reply ext/nnstreamer/extra/meson.build:37]: You're right. Although gRPC uses protobuf by default, it does not require any mandatory IDL.
But, in this design, each tensor_src/sink can choose IDL from a property, so I decided to check both dependencies.

If you think it does not make sense, I will try to distinguish dependencies.

File: ext/nnstreamer/extra/nnstreamer_grpc_common.cc [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    case GRPC_IDL_PROTOBUF:
+      return new ServiceImplProtobuf (server, host, port);
+    case GRPC_IDL_FLATBUF:
+      return new ServiceImplFlatbuf (server, host, port);
+    default:
+      return NULL;
+  }
+}
+
+/** @brief constructure of NNStreamerRPC */
```
[the remark this answers is not in the document]
Comment (dongju-chae) [reply ext/nnstreamer/extra/nnstreamer_grpc_common.cc:43]: Thx, will be updated.

File: ext/nnstreamer/extra/nnstreamer_grpc_common.cc [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+/** @brief start server service */
+gboolean
+NNStreamerRPC::_start_server () {
+  std::string address (host_);
+
+  address += ":" + std::to_string (port_);
+
+  grpc::EnableDefaultHealthCheckService(true);
```
[the remark this answers is not in the document]
Comment (dongju-chae) [reply ext/nnstreamer/extra/nnstreamer_grpc_common.cc:117]: Thx, will be updated.

File: ext/nnstreamer/extra/nnstreamer_grpc_common.cc [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+{
+  g_return_if_fail (priv != NULL);
+
+  NNStreamerRPC * self = static_cast<NNStreamerRPC *> (priv);
+
+  self->setConfig (config);
+}
+
+/**
+ * @brief gRPC C++ wrapper to set tensors config
```
[the remark this answers is not in the document]
Comment (dongju-chae) [reply ext/nnstreamer/extra/nnstreamer_grpc_common.cc:217]: Thx, will be updated.

File: ext/nnstreamer/extra/meson.build (hunk 5 above)
Comment (myungjoo) [reply ext/nnstreamer/extra/meson.build:37]: Please write an issue of this. Let's do it later.

**Pull request merged**